src/schema/firefox-schemas-import.js

```diff
@@ -250,24 +253,108 @@ export function rewriteExtend(schemas, schemaId) {
   return { definitions, refs, types };
 }
 
+export function filterSchemas(schemas) {
+  return schemas.filter((schema) => {
+    return !ignoredSchemas.includes(schema.namespace);
+  });
+}
+
+/**
+ * Merge multiple schemas into one if they are properties of each other.
+ *
+ * Example:
+ *
+ *  [{ namespace: "privacy", permissions: ["privacy"] },
+ *   { namespace: "privacy.network", properties: { networkPredictionEnabled: {} } }]
+ *
+ *  becomes
+ *
+ *  [{ namespace: "privacy",
+ *     permissions: ["privacy"],
+ *     properties: {
+ *       network: {
+ *         properties: {
+ *           networkPredictionEnabled: {}
+ *  }}}}]
+ */
+export function foldSchemas(schemas) {
+  // Map the schemas by prefix.
+  const schemasByPrefix = {};
+  schemas.forEach((schema) => {
+    const [prefix, property, more] = schema.namespace.split('.', 3);
+    if (more) {
+      throw new Error('namespace may only have one level of nesting');
+    }
+    if (!(prefix in schemasByPrefix)) {
+      schemasByPrefix[prefix] = {};
+    }
+    let namespace = property ? property : '__base';
+    if (schemasByPrefix[prefix][namespace]) {
+      throw new Error('matching namespaces are not allowed');
+    } else {
+      schemasByPrefix[prefix][namespace] = schema;
+    }
+  });
+  // If there aren't any matching prefixes then there's no folding to do.
+  const hasMatchingPrefixes = Object.keys(schemasByPrefix).some((prefix) => {
+    const prefixedSchemas = schemasByPrefix[prefix];
+    // Continue if there are multiple properties (__base and something else)
+    // or there is one property that isn't __base.
+    return Object.keys(prefixedSchemas).length > 1
+      || !('__base' in prefixedSchemas);
+  });
+  if (!hasMatchingPrefixes) {
+    return schemas;
+  }
+
+  // There is folding to do, join the matching schemas.
+  const foldedSchemas = [];
+
+  // The order of the schemas will be maintained since they were inserted in
+  // the order of schemas.
+  Object.keys(schemasByPrefix).forEach((namespace) => {
+    const { __base = {}, ...nestedSchemas } = schemasByPrefix[namespace];
```

Comment:
I figured it would be best to call it something that wouldn't be used by a schema.

It could be `baseNamespace`, that does seem pretty unlikely.